flocker/node/agents/cinder.py

```diff
@@ -0,0 +1,237 @@
+
+from keystoneclient_rackspace.v2_0 import RackspaceAuth
+from keystoneclient.session import Session
+
+from cinderclient.client import Client
+
+from zope.interface import implementer, Interface
+
+from .blockdevice import IBlockDeviceAPI, BlockDeviceVolume
+
+# Rackspace public docs say "The minimum size for a Cloud Block Storage volume
+# is 50 GB for an SSD volume or 75GB for an SATA volume. The maximum volume
+# size is 1TB."
+# * http://www.rackspace.com/knowledge_center/product-faq/cloud-block-storage
+# Rackspace API agrees "u'{"badRequest": {"message": "Invalid input
+# received: \'size\' parameter must be between 75 and 1024", "code":
+# 400}}'"
+# Let's assume that we only support SATA volumes for now.
+# Eventually we'll validate size at configuration time based on backend limits.
+# See https://clusterhq.atlassian.net/browse/FLOC-1579
+RACKSPACE_MINIMUM_BLOCK_SIZE = int(GB(75).to_Byte().value)
+RACKSPACE_MAXIMUM_BLOCK_SIZE = int(TB(1).to_Byte().value)
+
+# The key name used for identifying the Flocker cluster_id in the metadata for
+# a volume.
+CLUSTER_ID_LABEL = u'flocker-cluster-id'
+
+# The key name used for identifying the Flocker dataset_id in the metadata for
+# a volume.
+DATASET_ID_LABEL = u'flocker-dataset-id'
+
+# The Rackspace authentication endpoint
+# See http://docs.rackspace.com/cbs/api/v1.0/cbs-devguide/content/Authentication-d1e647.html # noqa
+RACKSPACE_AUTH_URL = "https://identity.api.rackspacecloud.com/v2.0"
+
+
+class ICinderVolumeManager(Interface):
+    """
+    The parts of ``cinderclient.v1.volumes.VolumeManager`` that we use.
+    See: https://github.com/openstack/python-cinderclient/blob/master/cinderclient/v1/volumes.py#L135 # noqa
+    """
+    def create(size, metadata=None):
+        """
+        Creates a volume.
+
+        :param size: Size of volume in GB
+        :param metadata: Optional metadata to set on volume creation
+        :rtype: :class:`Volume`
+        """
+
+    def list():
+        """
+        Lists all volumes.
+
+        :rtype: list of :class:`Volume`
+        """
+
+    def set_metadata(volume, metadata):
+        """
+        Update/Set a volumes metadata.
+
+        :param volume: The :class:`Volume`.
+        :param metadata: A list of keys to be set.
+        """
+
+
+def wait_for_volume(volume_manager, expected_volume):
+    """
+    Wait for a ``Volume`` with the same ``id`` as ``expected_volume`` to be
+    listed and to have a ``status`` value of ``available``.
+
+    :param ICinderVolumeManager volume_manager: An API for listing volumes.
+    :param Volume expected_volume: The ``Volume`` to wait for.
+    :returns: The listed ``Volume`` that matches ``new_volume``.
+    """
+    while True:
+        for listed_volume in volume_manager.list():
+            if listed_volume.id == expected_volume.id:
+                if listed_volume.status == 'available':
+                    return listed_volume
+
+
+@implementer(IBlockDeviceAPI)
+class CinderBlockDeviceAPI(object):
```

Comment:
Perhaps this should be a PRecord. It'd be nice to check the type of the supplied cluster_id and maybe check that the supplied volume_manager provides ICinderVolumeManager.